### PR TITLE
Clean up undefined behaviour

### LINKED ISSUE #276: Undefined behaviour
When enabling clang's UndefinedBehaviorSanitizer in liboqs, the follow warnings were detected in code copied from PQClean:

```
pqclean_kyber512_clean/reduce.c:21:21: runtime error: signed integer overflow: 4754970 * 62209 cannot be represented in type 'int'
pqclean_ntruhrss701_clean/sample.c:40:31: runtime error: signed integer overflow: 65535 * 65535 cannot be represented in type 'int'
pqclean_lightsaber_clean/poly_mul.c:40:106: runtime error: signed integer overflow: 65266 * 40970 cannot be represented in type 'int'
pqclean_lightsaber_clean/poly_mul.c:50:106: runtime error: signed integer overflow: 49144 * 49758 cannot be represented in type 'int'
pqclean_lightsaber_clean/poly_mul.c:56:44: runtime error: signed integer overflow: 59997 * 49156 cannot be represented in type 'int'
pqclean_lightsaber_clean/poly_mul.c:65:102: runtime error: signed integer overflow: 49147 * 46020 cannot be represented in type 'int'
pqclean_lightsaber_clean/poly_mul.c:45:44: runtime error: signed integer overflow: 65522 * 35511 cannot be represented in type 'int'
pqclean_lightsaber_clean/poly_mul.c:61:102: runtime error: signed integer overflow: 57352 * 49488 cannot be represented in type 'int'
pqclean_lightsaber_clean/poly_mul.c:69:48: runtime error: signed integer overflow: 57352 * 52037 cannot be represented in type 'int'
pqclean_lightsaber_clean/poly_mul.c:52:106: runtime error: signed integer overflow: 49142 * 57039 cannot be represented in type 'int'
pqclean_lightsaber_clean/poly_mul.c:41:106: runtime error: signed integer overflow: 39073 * 57320 cannot be represented in type 'int'
pqclean_lightsaber_clean/poly_mul.c:202:33: runtime error: signed integer overflow: 66146 * 36409 cannot be represented in type 'int'
pqclean_lightsaber_clean/poly_mul.c:204:31: runtime error: signed integer overflow: 1842504 * 61167 cannot be represented in type 'int'
pqclean_lightsaber_clean/poly_mul.c:200:33: runtime error: signed integer overflow: -433016 * 43691 cannot be represented in type 'int'
```

Same error(s) in Kyber{768,1024,512-90s,768-90s,1024-90s}, Saber, FireSaber.

All the errors seem to in places where some `int16_t` values are multiplied and overflow.

### PR DESCRIPTION
**Pull request by thomwiggers**
This PR:

* Fixes the undefined behaviour tests
* Enables the `-fsanitize` tests in CI (only on amd64)
* Fixes a bunch of UB, as described below.
 
Fix signed-integer overflow by doing a 64-bit multiplication and then casting it down.

* [x] Fix UB in Kybers
* [x] Fix UB in NTRU-HTRSS701
* [x] Fix UB in lightsaber
* [x] Fix UB in Saber
* [x] Fix UB in FireSaber

Closes #276

Alignment problems in Frodo (``crypto_kem_dec`` ``uint16_t`` use of ``uint8_t`` pointer):

* [x] Fix UB in Frodo640aes
  * [x] clean
  * [x] opt
* [x] Fix UB in Frodo640shake
  * [x] clean
  * [x] opt
* [x] Fix UB in Frodo976aes
  * [x] clean
  * [x] opt
* [x] Fix UB in Frodo976shake
  * [x] clean
  * [x] opt
* [x] Fix UB in Frodo1344aes
  * [x] clean
  * [x] opt
* [x] Fix UB in Frodo1344shake
  * [x] clean
  * [x] opt

Shift by width of int:

* [x] McEliece348864f vec
* [x] McEliece460896f vec

Shift left of signed int? in qTESLA:

* [x] qTESLA-p-I
* [x] qTESLA-p-III

### CODE REVIEW

**Comment by thomwiggers** [comment]
I'm just hoping that this works, and I don't have to do even more casting.

**Comment by thomwiggers** [comment]
... Of course, actually I should be writing this:

```c
int16_t result = (uint16_t)((uint32_t)a * (uint32_t)b);
```

And probably C won't infer that if I do it on the outside...

File: crypto_kem/ntruhrss701/clean/sample.c [hunk 1]
```diff
@@ -37,14 +37,14 @@ void PQCLEAN_NTRUHRSS701_CLEAN_sample_iid_plus(poly *r, const unsigned char unif
 
     /* s = <x*r, r>.  (r[n-1] = 0) */
     for (i = 0; i < NTRU_N - 1; i++) {
-        s += r->coeffs[i + 1] * r->coeffs[i];
+        s += (uint16_t)((uint32_t)r->coeffs[i + 1] * (uint32_t)r->coeffs[i]);
```
Comment (dstebila) [inline crypto_kem/ntruhrss701/clean/sample.c]: Did you want to use the `OVERFLOWING_MUL` macro here for consistency with the other implementations?

Comment (thomwiggers) [reply crypto_kem/ntruhrss701/clean/sample.c]: In SABER, it occurred so often that I thought it's more clear if I add a macro. Here it's only in two places – I don't think a macro makes it much more clear.

**Pull request merged**